File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

const solution = (input) => {
  const [n, m] = input[0].split(" ").map((v) => Number(v));
  const stair = Array(101).fill(0);
  const snake = Array(101).fill(0);
  const visited = Array(101).fill(Infinity);

  for (let i = 1; i <= n; i++) {
    const [u, v] = input[i].split(" ").map((v) => Number(v));
    stair[u] = v;
  }

  for (let i = n + 1; i < input.length; i++) {
    const [u, v] = input[i].split(" ").map((v) => Number(v));
    snake[u] = v;
  }

  const queue = [[1, 0]];
  visited[1] = 0;

  let answer = Infinity;

  // 도착지점의 마킹된 숫자보다 지금 숫자가 크면 안가는걸로
  while (queue.length) {
    const [position, cnt] = queue.shift();
    if (cnt >= answer) continue;

    if (position === 100) {
      answer = Math.min(answer, cnt);
    }

    if (stair[position] !== 0 && visited[stair[position]] >= cnt) {
      queue.push([stair[position], cnt]);
      visited[stair[position]] = cnt;
      visited[position] = cnt;
    }

    if (snake[position] !== 0 && visited[snake[position]] >= cnt) {
      queue.push([snake[position], cnt]);
      visited[snake[position]] = cnt;
      visited[position] = cnt;
    }

    for (let i = 1; i <= 6; i++) {
      const next = position + i;

      if (next > 100 || visited[next] <= cnt + 1) continue;
      visited[next] = cnt + 1;

      if (stair[next] !== 0) {
        queue.push([stair[next], cnt + 1]);
        visited[stair[next]] = cnt + 1;
        continue;
      }

      if (snake[next] !== 0) {
        queue.push([snake[next], cnt + 1]);
        visited[snake[next]] = cnt + 1;
        continue;
      }

      queue.push([next, cnt + 1]);
    }
  }

  console.log(answer);
};

solution(input);

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
from collections import deque

stairN, snakeN = list(map(int, input().split()))
stairPos = []
stairInfo = {}
snakePos = []
snakeInfo = {}

for i in range(stairN):
  s, e = map(int, input().split())
  stairPos.append(s)
  stairInfo[s] = e

for i in range(snakeN):
  s, e = map(int, input().split())
  snakePos.append(s)
  snakeInfo[s] = e

queue = deque()
queue.append([1, 0])
visited = [float('inf')]*101
visited[1] = 0
answer = float('inf')

while len(queue):
  pos, cnt = queue.popleft() 
  if pos == 100:
    answer = min(answer, cnt)

  if pos in stairPos and cnt <= visited[stairInfo[pos]]:
    queue.append([stairInfo[pos], cnt])
    visited[stairInfo[pos]] = cnt
    continue
  
  if pos in snakePos and cnt <= visited[snakeInfo[pos]]:
    queue.append([snakeInfo[pos], cnt])
    visited[snakeInfo[pos]] = cnt
    continue
  
  for i in range(1, 7):
    if i + pos <= 100 and cnt + 1 < visited[i+pos] and pos not in snakePos and pos not in stairPos:
      queue.append([i+pos, cnt + 1])
      visited[i+pos] = cnt + 1

print(answer)